hudi-utilities/src/main/java/org/apache/hudi/utilities/keygen/MultiFormatTimestampBasedKeyGenerator.java

```diff
@@ -0,0 +1,183 @@
+
+    this.timestampType = TimestampType.valueOf(config.getString(Config.TIMESTAMP_TYPE_FIELD_PROP));
+    this.outputDateFormat = config.getString(Config.TIMESTAMP_OUTPUT_DATE_FORMAT_PROP);
+    this.configInputDateFormatList = config.getString(Config.TIMESTAMP_INPUT_DATE_FORMAT_LIST_PROP, "");
+
+    String inputDateFormatDelimiter = this.config.getString(Config.TIMESTAMP_INPUT_DATE_FORMAT_LIST_DELIMETER_REGEX_PROP, ",").trim();
+    inputDateFormatDelimiter = inputDateFormatDelimiter.isEmpty() ? "," : inputDateFormatDelimiter;
+    this.configInputDateFormatDelimiter = inputDateFormatDelimiter;
+
+    if (inputTimeZone != null && !inputTimeZone.trim().isEmpty()) {
+      this.inputDateTimeZone = DateTimeZone.forTimeZone(TimeZone.getTimeZone(config.getString(Config.TIMESTAMP_INPUT_TIMEZONE_FORMAT_PROP, "")));
+    } else {
+      this.inputDateTimeZone = null;
+    }
+    if (outputTimeZone != null && !outputTimeZone.trim().isEmpty()) {
+      this.outputDateTimeZone = DateTimeZone.forTimeZone(TimeZone.getTimeZone(config.getString(Config.TIMESTAMP_OUTPUT_TIMEZONE_FORMAT_PROP, "")));
+    } else {
+      this.outputDateTimeZone = null;
+    }
+
+    if (timestampType == TimestampType.DATE_STRING || timestampType == TimestampType.MIXED) {
+      DataSourceUtils.checkRequiredProperties(config,
+          Collections.singletonList(Config.TIMESTAMP_INPUT_DATE_FORMAT_LIST_PROP));
+    }
+  }
+
+  private DateTimeFormatter getInputDateFormatter() {
+    if (this.configInputDateFormatList.isEmpty()) {
+      throw new IllegalArgumentException(Config.TIMESTAMP_INPUT_DATE_FORMAT_LIST_PROP + " configuration is required");
+    }
+
+    DateTimeFormatter formatter = new DateTimeFormatterBuilder()
+            .append(
+                    null,
+                    Arrays.asList(
+                            this.configInputDateFormatList.split(this.configInputDateFormatDelimiter))
+                            .stream()
+                            .map(String::trim)
+                            .map(DateTimeFormat::forPattern)
+                            .map(DateTimeFormatter::getParser)
+                            .toArray(DateTimeParser[]::new))
+            .toFormatter();
+    if (this.inputDateTimeZone != null) {
+      formatter = formatter.withZone(this.inputDateTimeZone);
+    } else {
+      formatter = formatter.withOffsetParsed();
+    }
+
+    return formatter;
+  }
+
+  @Override
+  public HoodieKey getKey(GenericRecord record) {
+    Object partitionVal = DataSourceUtils.getNestedFieldVal(record, partitionPathField, true);
+    if (partitionVal == null) {
+      partitionVal = 1L;
+    }
+
+    DateTimeFormatter inputFormatter = this.getInputDateFormatter();
+    DateTimeFormatter partitionFormatter = DateTimeFormat.forPattern(outputDateFormat);
+    if (this.outputDateTimeZone != null) {
+      partitionFormatter = partitionFormatter.withZone(outputDateTimeZone);
+    }
+
+    try {
+      long unixTime;
```

Comment:
wonder if we can reuse some code by having this class extend the TimestampBasedKeyGenerator..

Reply:
To extend the other class, we'd be bringing in instance variables that we'd not be leveraging in the new class - the Joda time vs SimpleDateFormat classes (per my previous reply above).

Reply:
How about refactoring the code in TimestampBasedKeyGenerarator itself to use Joda time there instead of SimpleDateFormat? We just need to be careful to support all the previous configs and the previous flow. Did you consider this option @allenerb ?

Reply:
In my opinion also, having 2 classes for timestamp based key generators can get confusing for users. :)

Reply:
Yup, that's fair - I don't mind refactoring this into the original class.  Shouldn't be difficult to make it backward compatible.